mineral-ui/mineral-ui
# Simplify package import paths

### Description

* Refactor packages to use `dist` and `lib` directories, `dist` for built files and `lib` for transpiled src.
* Flatten Icon component directory structure
* gitignore generated icon files
* Update documentation

**Questions/Comments:**
* Do `.flow.js` files need to live in `dist`?  It seems more intuitive that they would go in `lib`.
* There is a PR for adding flat package support to Lerna, which would be super nice, but I have a feel that it will not be accepted, based on previous conversations in that repo.

### Motivation and context

Please see #213 for background information

closes #213 

### How to test

1. Execute `npm install && npm run build && npm test` in the mineral project
2. Test built packages in another project
    * Setup another js app, perhaps using create-react-app or just another project that you have locally.
    * Install the mineral peer dependencies
    * Install the respective local mineral packages that you plan to test.  I used the following.
        * npm i -D path-to-mineral-ui/packages/component-utils
        * npm i -D path-to-mineral-ui/packages/icon
    * Import the respective components and use them in your app
    * Build your project
    * Note that the new import path works properly
    * Consider using different style import paths and noting how it affects your bundle size

I tested by following the above approach, using a local/internal project, called Sombrero.  Here are the results.

* New "lib" import path works properly
* Import syntax tests and resultant bundle sizes
    * named-export-from-index - `import { IconHelp } from @mineral-ui/icon` - 7.07 MB
    * default-export-from-index - `import Icon from @mineral-ui/icon` - 7.07 MB
    * named-export-from-lib - `import IconHelp from @mineral-ui/icon/lib/IconHelp` - 3.98 MB
    * default-export-from-lib - `import Icon from '@mineral-ui/icon/lib/Icon'` - 3.98 MB
* In both cases, importing directly from the lib provides the smallest bundle size, without relying on consumer tree-shaking.

### Types of changes

- Breaking change (fix or feature that would cause existing functionality to change)

Changes package distribution structure, thus could affect consumers that were reliant on the previous structure.

### Checklist


The only documentation that needed updating was for the Icon component.

### How does this PR make you feel?

![package](https://media.giphy.com/media/l0HlNOUe47hoQa3OE/giphy.gif)
